File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/NeedsWriteLockException.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import net.java.truecommons.shed._

/** Indicates that an operation needs to get write locked before it can get
  * retried.
  *
  * @author Christian Schlichtherle
  */
private final class NeedsWriteLockException private() extends ControlFlowException(false)

private object NeedsWriteLockException {

  import ControlFlowException._

  private[this] val instance = new NeedsWriteLockException

  def apply(): NeedsWriteLockException = {
    if (isTraceable) {
      new NeedsWriteLockException
    } else {
      instance
    }
  }
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/java/net/java/truevfs/kernel/impl/NeedsWriteLockException.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import net.java.truecommons.shed.ControlFlowException;

/**
 * Indicates that an operation needs to get write locked before it can get retried.
 *
 * @author Christian Schlichtherle
 */
final class NeedsWriteLockException extends ControlFlowException {

    private static final long serialVersionUID = 0;
    private static final NeedsWriteLockException instance = new NeedsWriteLockException();

    private NeedsWriteLockException() {
        super(false);
    }

    static NeedsWriteLockException apply() {
        return isTraceable() ? new NeedsWriteLockException() : instance;
    }
}
